### PR TITLE
SchemaV5: Refactor so two different editor service implementations can exist

### PR DESCRIPTION
**Pull request by ryan-rushton**
## Overview

This PR does some necessary refactoring so that we can have concurrent editor services available and manage them. I will list all the things done here, but that main one is that I have split the `DprintService` up into a few different classes/services. These are
- EditorService: An interface to define the API we use for all implementations
- EditorServiceV4: the current implementation of communication model with the underlying dprint `editor-service`, this is an IntelliJ service
- EditorServiceManager: an IntelliJ service to orchestrate creation of the correct implementation of the `EditorService` and expose it to the `FormatterService`
- EditorProcess: a class to facilitate the lower level communication with the editor service (read/write strings and ints), as well as to create/destroy the underlying system process

I have also
- Added the `RestartAction`, this should be similar to the VS Code one that restarts the underlying dprint `editor-service` and should be useful for changing config and restarting after errors
- Pulled out the `FormatResult` class so it can be reused across implementations
- Created the `ProcessUnavailableException` so we don't have to handle so many nulls

### CODE REVIEW

File: src/main/kotlin/com/dprint/services/editorservice/EditorProcess.kt [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+            else -> LOGGER.info(Bundle.message("editor.service.starting.working.dir", executablePath, configPath))
+        }
+
+        return commandLine.createProcess()
+    }
+
+    private fun getProcess(): Process {
+        return process
+            ?: throw ProcessUnavailableException(Bundle.message("editor.process.cannot.get.editor.service.process"))
+    }
```
Comment (ryan-rushton) [inline src/main/kotlin/com/dprint/services/editorservice/EditorProcess.kt:91]: This is the only new thing in this file. Everything else is copy paste from the old `DprintService`

File: src/main/kotlin/com/dprint/services/editorservice/FormatResult.kt [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+    /**
+     * The results of the formatting if successful.
+     */
+    var formattedContent: String? = null
+
+    /**
+     * The error message if formatting was not successful. This can come from custom messages, stderr or stdin.
+     */
+    var error: String? = null
+}
```
Comment (ryan-rushton) [inline src/main/kotlin/com/dprint/services/editorservice/FormatResult.kt:16]: Just extracted out of the old `DprintService`

File: src/main/kotlin/com/dprint/services/editorservice/ProcessUnavailableException.kt [hunk 3]
```diff
@@ -0,0 +1,8 @@
+package com.dprint.services.editorservice
+
+class ProcessUnavailableException: Exception {
+    constructor() : super()
+    constructor(message: String) : super(message)
+    constructor(message: String, cause: Throwable) : super(message, cause)
+    constructor(cause: Throwable) : super(cause)
+}
```
Comment (ryan-rushton) [inline src/main/kotlin/com/dprint/services/editorservice/ProcessUnavailableException.kt:8]: TODO: Pull this out into an exceptions package

File: src/main/kotlin/com/dprint/services/editorservice/v4/EditorServiceV4.kt [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+                    2 -> result.error = editorProcess.readString()
+                }
+
+                editorProcess.readAndAssertSuccess()
+            }
+        } catch (e: ProcessUnavailableException) {
+            val message = Bundle.message("editor.service.unable.to.determine.if.can.format", filePath)
+            notificationService.notifyOfFormatFailure(message)
+            LOGGER.info(message, e)
+        }
```
Comment (ryan-rushton) [inline src/main/kotlin/com/dprint/services/editorservice/v4/EditorServiceV4.kt:104]: Ditto about the try/catch

File: src/main/kotlin/com/dprint/services/editorservice/EditorServiceManager.kt [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+private const val SCHEMA_V4 = 4
+
+@Service
+class EditorServiceManager(private val project: Project) {
+    private var editorService: EditorService? = null
+    private val notificationService = project.service<NotificationService>()
+
+    init {
+        maybeInitialiseEditorService()
+    }
```
Comment (ryan-rushton) [inline src/main/kotlin/com/dprint/services/editorservice/EditorServiceManager.kt:28]: TODO: This is throwing "Synchronous execution on EDT" errors, we will need to defer this to the `maybeGetEditorService` method and ensure it is not called on the EDT thread.

**Review by dsherret — APPROVED** [review]
LGTM!

**Comment by ryan-rushton** [comment]
Merging this so I can start on the actual implementation

**Pull request merged**